input-events/input-events-typing.html

```diff
@@ -1,82 +1,311 @@
+        assert_equals(inputEvent.inputType, 'deleteContentForward');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing DELETE with no pre-existing content');
+
+    promise_test(async () => {
+        resetEditor('<p>Preexisting <i id="caret">c</i>ontent</p>');
+
+        await test_driver.click(document.querySelector('#caret'));
+        await test_driver.send_keys(document.getElementById('rich'), "\uE003"); // Back Space
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'deleteContentBackward');
+        assert_equals(inputEvent.inputType, 'deleteContentBackward');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing BACK_SPACE with pre-existing content');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await test_driver.send_keys(document.getElementById('rich'), "\uE003"); // Back Space
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'deleteContentBackward');
+        assert_equals(inputEvent.inputType, 'deleteContentBackward');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing BACK_SPACE with no pre-existing content');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await test_driver.send_keys(document.getElementById('rich'), "hello");
+
+        // Decide whether to use  Key.COMMAND (mac) or Key.CONTROL (everything else)
+        const modifierKey = navigator.userAgent.toLowerCase().includes('mac') ? '\u2318' : '\uE009';
+
+        // Undo
+        await new test_driver.Actions()
+            .keyDown(modifierKey)
+            .keyDown('z')
+            .keyUp('z')
+            .keyUp(modifierKey)
+            .send();
+        // Redo
+        await new test_driver.Actions()
+            .keyDown(modifierKey)
+            .keyDown('\uE008') // Shift
+            .keyDown('z')
+            .keyUp('z')
+            .keyUp('\uE008')
+            .keyUp(modifierKey)
+            .send();
+
+        // Ignore the initial typing of 'hello'
+        inputEventsLog = inputEventsLog.slice(10);
```

Comment:
nit: maybe make a new variable here? that way, if the test fails, the person that needs to debug it can more easily compare what is in the original `inputEventsLog` to whatever is in the new variable.

Reply:
done